miniapps/solvers/bramble_pasciak.hpp

```diff
@@ -0,0 +1,180 @@
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//          ----------------------------------------------------------
+//               Bramble-Pasciak preconditioning for Darcy problem
+//          ----------------------------------------------------------
+//
+// Main idea is to transform the block system
+//                 Ax = [ M  B^T ] [u] = [f] = b
+//                      [ B   0  ] [p] = [g]
+//     where:
+//        M = \int_\Omega (k u_h) \cdot v_h dx,
+//        B = -\int_\Omega (div_h u_h) q_h dx,
+//        u_h, v_h \in R_h (Raviart-Thomas finite element space),
+//        q_h \in W_h (piecewise discontinuous polynomials),
+// with a block transformation of the form X = A*N - Id
+//                  X = [ M*invQ - Id    0   ]
+//                      [     B*invQ    -Id  ]
+// where N is defined by
+//                  N = [ invQ    0 ]
+//                      [   0     0 ]
+// and Q is constructed such that Q and M-Q are both s.p.d..
+//
+// The solution x is then obtained by solving XAx = Xb with PCG as XA is s.p.d.
+//
+// The codes allows the user to provide such Q, or to construct it from the
+// element matrices M_T. Moreover, the user can provide a block preconditioner
+//                  P = [ M_0    0  ]
+//                      [  0    M_1 ]
+// for the transformed system XA.
+//
+// The code also allows the user to use BPCG, which is a special implementation
+// of the PCG iteration with the particular preconditioner H, defined as
+//                  H = [ M - Q    0  ]
+//                      [  0      M_1 ]
+// BPCG is efficient as it avoids the direct application of invH and X.
+
+#ifndef MFEM_BP_SOLVER_HPP
+#define MFEM_BP_SOLVER_HPP
+
+#include "darcy_solver.hpp"
+#include <memory>
+
+namespace mfem
+{
+namespace blocksolvers
+{
+
+/// Parameters for the BramblePasciakSolver method
+struct BPSParameters : IterSolveParameters
+{
+   bool use_bpcg = true;   // whether to use BPCG
+   double q_scaling = 0.5; // scaling (> 0 and < 1) of the Q preconditioner
+};
+
+/// Bramble-Pasciak Conjugate Gradient
+class BPCGSolver : public IterativeSolver
+{
+protected:
+   mutable Vector r, p, g, t, r_bar, r_red, g_red;
+   /// Remaining required operators
+   /*  Operator list
+    *  From IterativeSolver:
+    *  *oper  -> A  = [M, Bt; B, 0]
+    *  From this class:
+    *  *iprec -> N  = diag(M0, 0)
+    *  *pprec -> P' = diag(M0, M1) * [Id, 0; B*M0, -Id]
+    */
+   const Operator *iprec, *pprec;
+   void UpdateVectors();
+
+public:
+   BPCGSolver() { }
+   BPCGSolver(const Operator &ipc, const Operator &ppc) { pprec = &ppc; iprec = &ipc; }
+
+#ifdef MFEM_USE_MPI
+   BPCGSolver(MPI_Comm comm_) : IterativeSolver(comm_) { }
```

Comment:
Do we need the default constructor? (And similarly for the constructor that takes on the MPI communicator). If they are not used, I think it's better not to have them (i.e. to help ensure that the constructor object is in a valid state).